### PR TITLE
[v6] Add location prop to Routes and RouteOptions to useRoutes

### LINKED ISSUE #7117: Switch "location" prop in v6
Hey guys!

Amazing work on v6. Just spent the last few hours toying around and converting a few things. Made so many things so much easier!

Two things as I was messing around and looking through the code:

1. The <Link> component's "to" prop seems different? Instead passing an object like this
`props={pathname:  '/home', state: {cool: true}} <Link to={props}>Link</Link>`
I have to do this:
         `props={to:  '/home', state: {cool: true}} <Link {...props}>Link</Link>`

2. It seems like the "location" prop from the Switch component is gone with Routes. If so, how would I implement a pattern like this?
https://reacttraining.com/react-router/web/example/modal-gallery

### PR DESCRIPTION
**Pull request by smvv**
I cleaned the commits of #7298 to just these two. This should be easier to merge.

Fixes #7117 and #7297.

@MeiKatz @timdorr @mjackson @ryanflorence

### CODE REVIEW

**Comment by MeiKatz** [comment]
@smvv Thank you :) btw: are you a RR maintainer?

**Comment by smvv** [comment]
no, just a user that would like this feature as well!

**Comment by MeiKatz** [comment]
I am just surprised that you could edit my pull request.

File: packages/react-router/index.tsx [hunk 1]
```diff
@@ -608,12 +611,13 @@ function useRoutes_(
...
-    location,
-    routes,
-    basename
-  ]);
+  let currentLocation = useLocation() as Location;
+  let usedLocation = location || currentLocation;
+
+  let matches = React.useMemo(
+    () => matchRoutes(routes, usedLocation, basename),
+    [location, routes, basename]
```
Comment (kdembler) [inline packages/react-router/index.tsx:619]: `useMemo` dependencies weren't upgraded here with `usedLocation`

**Comment by MeiKatz** [comment]
@chaance To be honest: I am very disappointed about the way thing are going right now. Yes, it's great that things are starting again and ReactRouter v6 seems to point to an end and might be out of beta in the near future. But there are people like me who wanted to contribute changes and waited and waited and waited in the hope that they got merged. And now our pull requests are ignored and somebody else adds them on their own. It's kind of dishonoring the work of them. Sorry, but I am not really pleased by the things that happening right now.

**Comment by tconroy** [comment]
As someone who has also been following this issue/PR for a long time, I totally understand your frustration @MeiKatz  - I would be frustrated too. But in this case, it seems like an honest mistake with no ill intent. I assume since the other PR was open more recently, it surfaced higher in the review queue. I do think you and the other contribs should receive collaborative credit for this solution, though!

Appreciate everyone's work in getting this fixed,  I know my team appreciates it!

**Comment by chaance** [comment]
@MeiKatz I hear you, and I apologize and assure there was no ill-intent. Last week was my first on the job and we wanted to get lots of important bug fixes out since folks have been waiting for so long on some of these. I moved a bit too fast in this case. 🙇‍♂️

That said, the PR we merged ended up being reverted anyway (which I'm sure may feel just as frustrating!). @mjackson [started a new conversation](https://github.com/remix-run/react-router/issues/7946) on solving animations more directly in v6 that would eliminate the need for a `location` prop. We intend to have those APIs finalized very soon.

**Comment by timdorr** [comment]
I'm going to close this one out, not because I don't think this PR is good, but because there's been a lot of divergence and the revert on `dev` away from the `location` prop.

If the discussion does reveal a strong use case for including the `location` prop, then we should re-open this one, get it cleaned up, and merge it in. If that's where the API lands, I think it's only fair to give @MeiKatz some credit.

**Comment by wojtekmaj** [comment]
@timdorr I have shared my concerns in https://github.com/remix-run/react-router/issues/7946#issuecomment-900894584.

**Pull request closed without merging**